### PR TITLE
Feature/ssm agent add on

### PR DESCRIPTION
**Pull request by askulkarni2**
*Issue #, if available:* #93

*Description of changes:*

Installs SSM Agent on worker nodes using DaemonSet.

Before deployment:

```sh
aws ssm start-session --target i-06bbfe4fdeb6839d3

An error occurred (TargetNotConnected) when calling the StartSession operation: <instance-id> is not connected.
```

After deployment:
```sh
aws ssm start-session --target i-06bbfe4fdeb6839d3                             

Starting session with SessionId: kuapoorv-Isengard-0622c845c3ffc12e1
sh-4.2$ 
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/addons/ssm-agent.md
Comment (shapirov103) [inline docs/addons/ssm-agent.md:11]: Just a reminder for myself: we need a way to mark add-ons as incompatible with cluster providers, e.g. this one should be incompatible with Fargate / Bottlerocket.

File: docs/addons/ssm-agent.md
Comment (shapirov103) [inline docs/addons/ssm-agent.md:57]: assuming standard EKS AMI. socat may not be available for custom AMIs, we should call it out.

File: lib/addons/ssm-agent/index.ts
Comment (shapirov103) [inline lib/addons/ssm-agent/index.ts:39]: OMG, installs on every run? Isn't there an image with SSM pre-installed? This mutation won't work for any compliance programs, i am afraid. I am curious when eksctl is used with ssm = true, are they doing the same? (I realize that this is per documentation).

**Review by shapirov103 — COMMENTED** [review]
Overall, looks good and thank you for this. There are a couple of comments, one of them mainly research if there is an immutable image for ssm.

File: docs/addons/ssm-agent.md
Comment (askulkarni2) [reply docs/addons/ssm-agent.md:57]: Added callout in Limitations

File: lib/addons/ssm-agent/index.ts
Comment (askulkarni2) [reply lib/addons/ssm-agent/index.ts:39]: As per https://github.com/awslabs/amazon-eks-ami/issues/404 and https://github.com/awslabs/amazon-eks-ami/issues/127, SSM agent is not considered required for EKS and hence not added in the AMI.

Comment (shapirov103) [reply lib/addons/ssm-agent/index.ts:39]: I was thinking more along the lines of an AWS authored container image with SSM preinstalled. It would be nice to have it in AMI, but we also custom AMIs which won't have it.

Comment (shapirov103) [reply lib/addons/ssm-agent/index.ts:39]: A few things before we close it out:

1. Let's use https://github.com/jicowan/ssm-agent to build the image and upload it to a public ECR.
2. We need public ECR repo for a number of things including bootstrapping so let's create a task for that (@kcoleman731 wdyt?)
3. Let's upvote the following on GitHub: https://github.com/aws/amazon-ssm-agent/issues/135 (@parkand1 @youngjeong46 please join in).

Comment (askulkarni2) [reply lib/addons/ssm-agent/index.ts]: @shapirov103 added $1 from above comment. Will need $2 (cc: @kcoleman731)

Comment (shapirov103) [reply lib/addons/ssm-agent/index.ts]: Just making sure, as it stands it will work by pulling public image from your repository, correct? I am going to merge it right away to test if that is the case.

Comment (askulkarni2) [reply lib/addons/ssm-agent/index.ts]: That's correct. Any one can docker pull that image.

File: lib/addons/ssm-agent/ssm-install.sh
Comment (shapirov103) [inline lib/addons/ssm-agent/ssm-install.sh:3]: Can we point to an explicit version of the agent and tag the image with that version, so that we could have deterministic upgrades? The use case that I have in mind: we will upgrade the agent and publish it implicitly as latest. Customers will pick it up immediately so that in the existing cluster the SSP will be running a different version of the agent with the same SSP version. It is good when the agent works flawlessly, but it is great if one of the versions fails for any reason and we can't reproduce the state.

**Review by shapirov103 — COMMENTED** [review]
A couple of easily addressable comments. The version/tag one is important IMO. I will merge right after, I realized that I need a private cluster to really test it out.

File: docs/addons/ssm-agent.md
Comment (askulkarni2) [reply docs/addons/ssm-agent.md:72]: done.

**Comment by askulkarni2** [comment]
Found a `CrashLoopBackOff` issue. I am preparing a fix for it.

File: docs/addons/ssm-agent.md
Comment (kcoleman731) [reply docs/addons/ssm-agent.md:11]: Ticket here. https://github.com/aws-quickstart/quickstart-ssp-amazon-eks/issues/134

File: docs/addons/ssm-agent.md
Comment (kcoleman731) [inline docs/addons/ssm-agent.md:48]: @askulkarni2 during validation, I ran into an issue where I didn't have the `Session Manager plugin for the AWS CLI installed`. We should add that as a step here. 

https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with-install-plugin.html

**Review by kcoleman731 — APPROVED** [review]
@shapirov103 validated this works like a charm. Nice work @askulkarni2 

I left one comment re the Session Manager plugin needing to be installed. Please add that and we will be good to merge.

File: docs/addons/ssm-agent.md
Comment (askulkarni2) [reply docs/addons/ssm-agent.md:50]: addressed

File: lib/addons/ssm-agent/index.ts
Comment (askulkarni2) [inline lib/addons/ssm-agent/index.ts:32]: Added a pause container to avoid `CrashLoopBackoff` errors.

File: lib/addons/ssm-agent/ssm-install.sh
Comment (askulkarni2) [reply lib/addons/ssm-agent/ssm-install.sh:3]: addressed

**Review by kcoleman731 — APPROVED** [review]

**Pull request merged**
